### PR TITLE
Linkage monitor error to include dependency path to problematic artifacts

### PR DESCRIPTION
**Pull request by suztomo**
Toward #981 . Option 2: "Show Path to the Problematic Artifact" in the issue description. Not showing scope (yet).

With this PR, when Linkage Monitor shows error, it tells the location of the problematic Maven artifact in the dependency tree.

Example case to show gwt-user-2.8.2.jar:
```
suztomo@suxtomo24:~/cloud-opensource-java/linkage-monitor$ java -jar target/linkage-monitor-1.0.1-SNAPSHOT-all-deps.jar   com.google.cloud:libraries-bom
BOM Coordinates: com.google.cloud:libraries-bom:2.7.1-SNAPSHOT
Newly introduced problems:
...
Class org.objectweb.asm.commons.Method is not found
  referenced from com.google.web.bindery.requestfactory.server.RequestFactoryJarExtractor (gwt-user-2.8.2.jar)
Class com.google.gwt.dev.util.Name$SourceOrBinaryName is not found
  referenced from com.google.web.bindery.requestfactory.server.RequestFactoryJarExtractor (gwt-user-2.8.2.jar)
Class org.hibernate.validator.engine.ConstraintViolationImpl is not found
  referenced from org.hibernate.validator.engine.ConstraintViolationImpl_CustomFieldSerializer (gwt-user-2.8.2.jar)
Class org.hibernate.validator.engine.PathImpl is not found
  referenced from org.hibernate.validator.engine.PathImpl_CustomFieldSerializer (gwt-user-2.8.2.jar)
gwt-user-2.8.2.jar is at:
  com.google.cloud:google-cloud-firestore:1.28.1-SNAPSHOT / com.google.truth:truth:1.0 / com.google.gwt:gwt-user:2.8.2

Exception in thread "main" com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitorException: Found 278 new linkage errors
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.run(LinkageMonitor.java:137)
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.main(LinkageMonitor.java:87)

```

### CODE REVIEW

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java [hunk 1]
```diff
@@ -180,22 +183,29 @@ public void testBomSnapshot()
   @Test
   public void generateMessageForNewError() {
     Set<SymbolProblem> baselineProblems = ImmutableSet.of(classNotFoundProblem);
+    Path jar = Paths.get("aaa", "ccc-1.2.3.jar");
```
Comment (suztomo) [inline linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java]: This dummy JAR file is renamed from bbb to ccc to distinguish it from [the line 178's dummy data](https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/983/files#diff-36a547c579afc122ddf52d51a6199561L178).

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java [hunk 2]
```diff
@@ -180,22 +183,29 @@ public void testBomSnapshot()
...
+            snapshotProblems, baselineProblems, ImmutableListMultimap.of(jar, dependencyPath));
     assertEquals(
         "Newly introduced problem:\n"
             + "(bbb-1.2.3.jar) io.grpc.protobuf.ProtoUtils.marshaller's method"
             + " marshaller(com.google.protobuf.Message arg1) is not found\n"
-            + "  referenced from com.abc.AAA (bbb-1.2.3.jar)\n"
-            + "  referenced from com.abc.BBB (bbb-1.2.3.jar)\n",
+            + "  referenced from com.abc.AAA (ccc-1.2.3.jar)\n"
+            + "  referenced from com.abc.BBB (ccc-1.2.3.jar)\n"
+            + "ccc-1.2.3.jar is at:\n  foo:bar:1.0.0 / aaa:ccc:1.2.3\n",
```
Comment (suztomo) [inline linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java]: This PR does not enrich scope (yet).

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java [hunk 3]
```diff
@@ -180,22 +183,29 @@ public void testBomSnapshot()
...
-    String message = LinkageMonitor.messageForNewErrors(snapshotProblems, baselineProblems);
+    DependencyPath dependencyPath = new DependencyPath();
+    dependencyPath.add(new DefaultArtifact("foo:bar:1.0.0"));
+    dependencyPath.add(new DefaultArtifact("aaa:ccc:1.2.3"));
+    String message =
+        LinkageMonitor.messageForNewErrors(
+            snapshotProblems, baselineProblems, ImmutableListMultimap.of(jar, dependencyPath));
     assertEquals(
         "Newly introduced problem:\n"
             + "(bbb-1.2.3.jar) io.grpc.protobuf.ProtoUtils.marshaller's method"
```
Comment (elharo) [inline linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java]: is the method name (here "marshaller") being listed twice?

Comment (suztomo) [reply linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java]: Good catch.

Fixed the dummy data.

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java [hunk 4]
```diff
@@ -207,7 +217,7 @@ public void testGenerateMessageForFixedError() {
     assertEquals(
         "The following problems in the baseline no longer appear in the snapshot:\n"
```
Comment (elharo) [inline linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java]: don't we want problems in the snapshot that do not appear in the baseline?

Comment (suztomo) [reply linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitorTest.java]: Yes, we want that.

This test case testGenerateMessageForFixedError is to show the opposite: showing problems in the baseline that do not appear in the snapshot. Such problems are just recorded and they do not throw error.

**Pull request merged**